src/bit-fields/src/bit_range.rs

```diff
@@ -0,0 +1,1022 @@
+                let mut x = ", stringify!($y), "::new(1).unwrap();
+
+                assert_eq!(BitRangeMut::<_,0,1>(&mut x).get_max(),", stringify!($y), "::new(1).unwrap());
+                assert_eq!(BitRangeMut::<_,0,2>(&mut x).get_max(),", stringify!($y), "::new(3).unwrap());
+                assert_eq!(BitRangeMut::<_,0,3>(&mut x).get_max(),", stringify!($y), "::new(7).unwrap());
+                assert_eq!(BitRangeMut::<_,0,4>(&mut x).get_max(),", stringify!($y), "::new(15).unwrap());
+                assert_eq!(BitRangeMut::<_,0,5>(&mut x).get_max(),", stringify!($y), "::new(31).unwrap());
+                assert_eq!(BitRangeMut::<_,0,6>(&mut x).get_max(),", stringify!($y), "::new(63).unwrap());
+                assert_eq!(BitRangeMut::<_,0,7>(&mut x).get_max(),", stringify!($y), "::new(127).unwrap());
+                assert_eq!(BitRangeMut::<_,0,8>(&mut x).get_max(),", stringify!($y), "::new(255).unwrap());
+                ```
+            ")]
+            #[must_use]
+            #[inline]
+            pub const fn get_max(&self) -> $y {
+                Self::MAX
+            }
+            #[doc = concat!("
+                Convenience alias for [`Self::MASK`].
+                
+                ```
+                use bit_fields::BitRangeMut;
+                use std::num::", stringify!($y), ";
+                let mut x = ", stringify!($y), "::new(1).unwrap();
+
+                assert_eq!(BitRangeMut::<_,0,1>(&mut x).get_mask(),", stringify!($y), "::new(1).unwrap());
+                assert_eq!(BitRangeMut::<_,1,3>(&mut x).get_mask(),", stringify!($y), "::new(6).unwrap());
+                assert_eq!(BitRangeMut::<_,3,7>(&mut x).get_mask(),", stringify!($y), "::new(120).unwrap());
+                assert_eq!(BitRangeMut::<_,7,8>(&mut x).get_mask(),", stringify!($y), "::new(128).unwrap());
+                ```
+            ")]
+            #[must_use]
+            #[inline]
+            pub const fn get_mask(&self) -> $y {
+                Self::MASK
+            }
+        }
+    };
+}
+
+bit_mut_range!(u128, NonZeroU128, mask_u128, max_u128);
+bit_mut_range!(u64, NonZeroU64, mask_u64, max_u64);
+bit_mut_range!(u32, NonZeroU32, mask_u32, max_u32);
+bit_mut_range!(u16, NonZeroU16, mask_u16, max_u16);
+bit_mut_range!(u8, NonZeroU8, mask_u8, max_u8);
+
+/// Macro for defining `From` implementations on `BitRange` and `BitRangeMut`.
+macro_rules! bit_range_from {
+    ($x:ty, $y:ty, $bit_range: ident) => {
+        // `START < 8 * size_of::<$x>()` is always true so the right shift will not panic.
+        #[allow(clippy::integer_arithmetic, clippy::arithmetic_side_effects)]
+        impl<const START: u8, const END: u8> From<&$bit_range<'_, $x, START, END>> for $x {
+            #[inline]
+            fn from(this: &$bit_range<'_, $x, START, END>) -> Self {
+                let a = $bit_range::<'_, $x, START, END>::MASK & *this.0;
+                a >> START
+            }
+        }
```

Comment:
The comment about not panicing is wrong. The following compiles and panics:

```rs
#[test]
fn break_it() {
    let value = 8u8;
    let range = BitRange::<u8, 10, 12>(&value);
    dbg!(u8::from(&range));
}
```

Reply:
Generally, a lot of things panic in debug mode if bit ranges with `START > sizeof<T> * 8` are constructed. The NonZero* variants actually catch these because the compiler will complain about invalid constants being created (see above). Maybe we could always mark MASK and MAX as non-zero, to catch all of these at compile time?

Reply:
Added asserts to error at compile time with invalid values `START` and `END`.